File: MiniDOMAsteriskLib/CallManagers/Objects/Peers.vb
Imports System.Net.Sockets
Imports System.Threading
Imports System.Net
Imports minidom.CallManagers.Responses

Namespace CallManagers


    Public Class Peers
        Inherits System.Collections.ArrayList 'CSyncKeyCollection(Of Peer)

        Private m_Owner As AsteriskCallManager

        Public Sub New()
            DMDObject.IncreaseCounter(Me)
        End Sub

        Public Overrides ReadOnly Property IsSynchronized As Boolean
            Get
                Return True
            End Get
        End Property

        Public Sub New(ByVal owner As AsteriskCallManager)
            Me.New
            Me.m_Owner = owner
        End Sub

        'Protected Overrides Sub OnInsert(index As Integer, value As Object)
        '    If (Me.m_Owner IsNot Nothing) Then DirectCast(value, Peer).SetOwner(Me.m_Owner)
        '    MyBase.OnInsert(index, value)
        'End Sub

        Public Shadows Sub Add(ByVal value As Peer)
            If (Me.m_Owner IsNot Nothing) Then DirectCast(value, Peer).SetOwner(Me.m_Owner)
            MyBase.Add(value)
        End Sub

        Public Function GetItemByKey(ByVal name As String) As Peer
            SyncLock Me.SyncRoot
                For Each item As Peer In Me
                    If item.Name = name Then Return item
                Next
                Return Nothing
            End SyncLock
        End Function

        Protected Friend Sub UpdatePeerStatus(ByVal e As Events.PeerStatus)
            SyncLock Me.SyncRoot
                Dim peer As Peer = Me.GetItemByKey(e.Peer)
                If (peer Is Nothing) Then
                    peer = New Peer
                    peer.Name = e.Peer
                    Me.Add(peer) 'e.Peer) ', peer)
                End If
                peer.LastUpdated = Now
                peer.Status = e.PeerStatus
            End SyncLock
        End Sub

        Protected Overrides Sub Finalize()
            MyBase.Finalize()
            DMDObject.DecreaseCounter(Me)
        End Sub
    End Class

End Namespace

File: MiniDOMAsteriskLib/CallManagers/Objects/Peers.cs
using System.Collections;
using DMD;

namespace minidom.CallManagers
{
    public class Peers : ArrayList // CSyncKeyCollection(Of Peer)
    {
        private AsteriskCallManager m_Owner;

        public Peers()
        {
            DMDObject.IncreaseCounter(this);
        }

        public override bool IsSynchronized
        {
            get
            {
                return true;
            }
        }

        public Peers(AsteriskCallManager owner) : this()
        {
            m_Owner = owner;
        }

        // Protected Overrides Sub OnInsert(index As Integer, value As Object)
        // If (Me.m_Owner IsNot Nothing) Then DirectCast(value, Peer).SetOwner(Me.m_Owner)
        // MyBase.OnInsert(index, value)
        // End Sub

        public new void Add(Peer value)
        {
            if (m_Owner is object)
                value.SetOwner(m_Owner);
            base.Add(value);
        }

        public Peer GetItemByKey(string name)
        {
            lock (SyncRoot)
            {
                foreach (Peer item in this)
                {
                    if ((item.Name ?? "") == (name ?? ""))
                        return item;
                }

                return null;
            }
        }

        protected internal void UpdatePeerStatus(Events.PeerStatusType e)
        {
            lock (SyncRoot)
            {
                var peer = GetItemByKey(e.Peer);
                if (peer is null)
                {
                    peer = new Peer();
                    peer.Name = e.Peer;
                    Add(peer); // e.Peer) ', peer)
                }

                peer.LastUpdated = DMD.DateUtils.Now();
                peer.Status = e.PeerStatus;
            }
        }

        ~Peers()
        {
            DMDObject.DecreaseCounter(this);
        }
    }
}